File: go-honeypot/server/internal/honeypots/redis.go
package honeypots

import (
	"context"
	"errors"
	"net"
	"strconv"
	"strings"

	"github.com/example/honeypot/internal/config"
	"github.com/example/honeypot/internal/eventlog"
)

func NewRedis(cfg config.Service, store *eventlog.Store) *TCP {
	return NewTCP("redis", cfg, store, func(ctx context.Context, conn net.Conn, sessionID string, meta ConnMeta) {
		var buf strings.Builder
		tmp := make([]byte, 512)
		for {
			n, err := conn.Read(tmp)
			if err != nil {
				return
			}
			buf.Write(tmp[:n])
			cmds, remainder := parseResp(buf.String())
			buf.Reset()
			buf.WriteString(remainder)
			for _, parts := range cmds {
				upper := ""
				if len(parts) > 0 {
					upper = strings.ToUpper(parts[0])
				}
				evt := eventlog.Event{
					Service: "redis", Type: "command", SessionID: sessionID,
					RemoteIP: meta.RemoteIP, RemotePort: meta.RemotePort, LocalPort: meta.LocalPort,
					Command: strings.Join(parts, " "),
				}
				switch upper {
				case "AUTH":
					evt.Type = "login_attempt"
					if len(parts) > 2 {
						evt.Username = parts[1]
						evt.Password = parts[2]
					} else if len(parts) > 1 {
						evt.Username = "default"
						evt.Password = parts[1]
					}
					store.Log(evt)
					_, _ = conn.Write([]byte("-WRONGPASS invalid username-password pair or user is disabled.\r\n"))
				case "PING":
					store.Log(evt)
					_, _ = conn.Write([]byte("+PONG\r\n"))
				case "QUIT":
					store.Log(evt)
					_, _ = conn.Write([]byte("+OK\r\n"))
					return
				default:
					store.Log(evt)
					_, _ = conn.Write([]byte("-NOAUTH Authentication required.\r\n"))
				}
			}
			_ = ctx
		}
	})
}

// parseResp parses one or more RESP arrays (or inline commands) from `s`.
func parseResp(s string) ([][]string, string) {
	var out [][]string
	i := 0
	for i < len(s) {
		if s[i] == '*' {
			nl := strings.Index(s[i:], "\r\n")
			if nl < 0 {
				break
			}
			count, err := strconv.Atoi(s[i+1 : i+nl])
			if err != nil {
				break
			}
			off := i + nl + 2
			parts := make([]string, 0, count)
			ok := true
			for n := 0; n < count; n++ {
				if off >= len(s) || s[off] != '$' {
					ok = false
					break
				}
				nl2 := strings.Index(s[off:], "\r\n")
				if nl2 < 0 {
					ok = false
					break
				}
				strLen, err := strconv.Atoi(s[off+1 : off+nl2])
				if err != nil {
					ok = false
					break
				}
				off += nl2 + 2
				if off+strLen+2 > len(s) {
					ok = false
					break
				}
				parts = append(parts, s[off:off+strLen])
				off += strLen + 2
			}
			if !ok {
				break
			}
			out = append(out, parts)
			i = off
		} else {
			nl := strings.Index(s[i:], "\r\n")
			if nl < 0 {
				break
			}
			line := s[i : i+nl]
			fields := strings.Fields(line)
			if len(fields) > 0 {
				out = append(out, fields)
			}
			i += nl + 2
		}
	}
	return out, s[i:]
}

var _ = errors.New

File: src/honeypots/redis.js
'use strict';

const { TcpHoneypot } = require('./base');

class RedisHoneypot extends TcpHoneypot {
  constructor(cfg, logger) {
    super({ name: 'redis', port: cfg.port }, logger);
  }

  onConnection(socket, meta) {
    let buf = '';
    const send = (s) => { try { socket.write(s); } catch { /* noop */ } };

    socket.on('data', (data) => {
      buf += data.toString('utf8');
      const commands = parseRespCommands(buf);
      buf = commands.remainder;
      for (const cmd of commands.list) {
        const upper = cmd[0]?.toUpperCase() || '';
        this.logger.log({
          service: 'redis',
          type: upper === 'AUTH' ? 'login_attempt' : 'command',
          sessionId: meta.sessionId,
          remoteIp: meta.remoteIp,
          remotePort: meta.remotePort,
          localPort: meta.localPort,
          command: cmd.join(' '),
          username: upper === 'AUTH' ? (cmd.length > 2 ? cmd[1] : 'default') : null,
          password: upper === 'AUTH' ? (cmd.length > 2 ? cmd[2] : cmd[1]) : null
        });
        if (upper === 'AUTH') {
          send('-WRONGPASS invalid username-password pair or user is disabled.\r\n');
        } else if (upper === 'PING') {
          send('+PONG\r\n');
        } else if (upper === 'QUIT') {
          send('+OK\r\n');
          try { socket.end(); } catch { /* noop */ }
        } else {
          send('-NOAUTH Authentication required.\r\n');
        }
      }
    });
  }
}

function parseRespCommands(input) {
  const list = [];
  let i = 0;
  while (i < input.length) {
    const start = i;
    if (input[i] === '*') {
      const nl = input.indexOf('\r\n', i);
      if (nl < 0) break;
      const count = parseInt(input.slice(i + 1, nl), 10);
      i = nl + 2;
      const parts = [];
      let ok = true;
      for (let n = 0; n < count; n += 1) {
        if (input[i] !== '$') { ok = false; break; }
        const nl2 = input.indexOf('\r\n', i);
        if (nl2 < 0) { ok = false; break; }
        const len = parseInt(input.slice(i + 1, nl2), 10);
        i = nl2 + 2;
        if (input.length < i + len + 2) { ok = false; break; }
        parts.push(input.slice(i, i + len));
        i += len + 2;
      }
      if (!ok) { i = start; break; }
      list.push(parts);
    } else {
      // inline command
      const nl = input.indexOf('\r\n', i);
      if (nl < 0) break;
      const parts = input.slice(i, nl).split(/\s+/).filter(Boolean);
      if (parts.length) list.push(parts);
      i = nl + 2;
    }
  }
  return { list, remainder: input.slice(i) };
}

module.exports = { create: (cfg, logger) => new RedisHoneypot(cfg, logger) };
